File: meowlang/lexer.py
"""
Lexer pour MeowLang - Tokenisation du code source.
"""

import re
from typing import List, Optional, Tuple
from enum import Enum, auto


class TokenType(Enum):
    MIAOU = auto()
    MEOW = auto()
    
    ECRIRE = auto()
    DEMANDER = auto()
    TEXTE = auto()
    NOMBRE = auto()
    
    SI = auto()
    ALORS = auto()
    SINON = auto()
    SINON_SI = auto()
    
    REPETER = auto()
    FOIS = auto()
    TANT_QUE = auto()
    POUR_CHAQUE = auto()
    DANS = auto()
    COMPTEUR = auto()
    
    FONCTION = auto()
    RETOUR = auto()
    
    LISTE = auto()
    DICTIONNAIRE = auto()
    
    ESSAYER = auto()
    SAUF = auto()
    ERREUR = auto()
    
    IMPORTER = auto()
    
    MINUSCULE = auto()
    MAJUSCULE = auto()
    LONGUEUR = auto()
    REMPLACER = auto()
    CONTIENT = auto()
    
    ALEATOIRE = auto()
    SQRT = auto()
    ABS = auto()
    ROUND = auto()
    FLOOR = auto()
    CEIL = auto()
    
    OUVRIR = auto()
    LIRE = auto()
    FERMER = auto()
    
    ATTENDRE = auto()
    
    IDENTIFIER = auto()
    STRING = auto()
    NUMBER = auto()
    BOOLEAN = auto()
    
    PLUS = auto()
    MINUS = auto()
    MULTIPLY = auto()
    DIVIDE = auto()
    FLOOR_DIV = auto()
    MODULO = auto()
    POWER = auto()
    
    ASSIGN = auto()
    EQUAL = auto()
    NOT_EQUAL = auto()
    LESS_THAN = auto()
    GREATER_THAN = auto()
    LESS_EQUAL = auto()
    GREATER_EQUAL = auto()
    
    ET = auto()
    OU = auto()
    NON = auto()
    
    A = auto()
    
    COLON = auto()
    COMMA = auto()
    LPAREN = auto()
    RPAREN = auto()
    LBRACKET = auto()
    RBRACKET = auto()
    DOT = auto()
    
    NEWLINE = auto()
    INDENT = auto()
    DEDENT = auto()
    
    COMMENT = auto()
    EOF = auto()


class Token:
    def __init__(self, type_: TokenType, value: any, line: int, column: int):
        self.type = type_
        self.value = value
        self.line = line
        self.column = column
    
    def __repr__(self):
        return f"Token({self.type.name}, {self.value!r}, {self.line}:{self.column})"
    
    def __eq__(self, other):
        if not isinstance(other, Token):
            return False
        return (self.type == other.type and 
                self.value == other.value and 
                self.line == other.line and 
                self.column == other.column)


KEYWORDS = {
    'miaou': TokenType.MIAOU,
    'meow': TokenType.MEOW,
    'ecrire': TokenType.ECRIRE,
    'demander': TokenType.DEMANDER,
    'si': TokenType.SI,
    'alors': TokenType.ALORS,
    'sinon': TokenType.SINON,
    'repeter': TokenType.REPETER,
    'fois': TokenType.FOIS,
    'tant': TokenType.TANT_QUE,
    'que': TokenType.TANT_QUE,
    'pour': TokenType.POUR_CHAQUE,
    'chaque': TokenType.POUR_CHAQUE,
    'dans': TokenType.DANS,
    'compteur': TokenType.COMPTEUR,
    'fonction': TokenType.FONCTION,
    'retour': TokenType.RETOUR,
    'liste': TokenType.LISTE,
    'dictionnaire': TokenType.DICTIONNAIRE,
    'essayer': TokenType.ESSAYER,
    'sauf': TokenType.SAUF,
    'erreur': TokenType.ERREUR,
    'importer': TokenType.IMPORTER,
    'minuscule': TokenType.MINUSCULE,
    'majuscule': TokenType.MAJUSCULE,
    'longueur': TokenType.LONGUEUR,
    'remplacer': TokenType.REMPLACER,
    'contient': TokenType.CONTIENT,
    'aleatoire': TokenType.ALEATOIRE,
    'sqrt': TokenType.SQRT,
    'abs': TokenType.ABS,
    'round': TokenType.ROUND,
    'floor': TokenType.FLOOR,
    'ceil': TokenType.CEIL,
    'ouvrir': TokenType.OUVRIR,
    'lire': TokenType.LIRE,
    'fermer': TokenType.FERMER,
    'attendre': TokenType.ATTENDRE,
    'vrai': TokenType.BOOLEAN,
    'faux': TokenType.BOOLEAN,
    'et': TokenType.ET,
    'ou': TokenType.OU,
    'non': TokenType.NON,
}


class Lexer:
    def __init__(self, source: str, filename: str = "<script>"):
        self.source = source
        self.filename = filename
        self.lines = source.split('\n')
        self.pos = 0
        self.line = 1
        self.column = 1
        self.tokens: List[Token] = []
        self.indent_stack = [0]
        self.at_line_start = True
    
    def error(self, message: str, code: str = "E100"):
        from .errors import ErrorCatalog
        ErrorCatalog.raise_error(
            code=code,
            file=self.filename,
            line=self.line,
            column=self.column,
            instruction=self.current_char() if self.pos < len(self.source) else "",
            source_lines=self.lines
        )
    
    def current_char(self) -> Optional[str]:
        if self.pos >= len(self.source):
            return None
        return self.source[self.pos]
    
    def peek_char(self, offset: int = 1) -> Optional[str]:
        pos = self.pos + offset
        if pos >= len(self.source):
            return None
        return self.source[pos]
    
    def advance(self):
        if self.pos < len(self.source):
            if self.source[self.pos] == '\n':
                self.line += 1
                self.column = 1
                self.at_line_start = True
            else:
                self.column += 1
            self.pos += 1
    
    def skip_whitespace(self, skip_newlines: bool = True):
        while self.current_char() and self.current_char() in ' \t\r':
            self.advance()
        
        if skip_newlines and self.current_char() == '\n':
            self.advance()
            self.skip_whitespace(skip_newlines=True)
    
    def skip_comment(self):
        if self.current_char() == '#':
            while self.current_char() and self.current_char() != '\n':
                self.advance()
    
    def read_string(self) -> Tuple[str, int, int]:
        start_line = self.line
        start_column = self.column
        quote_char = self.current_char()
        self.advance()
        
        result = []
        while self.current_char() and self.current_char() != quote_char:
            if self.current_char() == '\\':
                self.advance()
                if self.current_char() in ('n', 't', 'r', '\\', '"', "'"):
                    escape_map = {'n': '\n', 't': '\t', 'r': '\r', '\\': '\\', '"': '"', "'": "'"}
                    result.append(escape_map.get(self.current_char(), self.current_char()))
                    self.advance()
                else:
                    result.append(self.current_char())
                    self.advance()
            else:
                result.append(self.current_char())
                self.advance()
        
        if not self.current_char():
            self.line = start_line
            self.column = start_column
            self.error("Guillemet de fermeture manquant", code="E101")
        
        self.advance()
        return ''.join(result), start_line, start_column
    
    def read_number(self) -> Tuple[TokenType, float, int, int]:
        start = self.pos
        start_line = self.line
        start_column = self.column
        has_dot = False
        
        while self.current_char() and (self.current_char().isdigit() or self.current_char() == '.'):
            if self.current_char() == '.':
                if has_dot or not self.peek_char() or not self.peek_char().isdigit():
                    break
                has_dot = True
            self.advance()
        
        number_str = self.source[start:self.pos]
        
        if has_dot:
            return TokenType.NUMBER, float(number_str), start_line, start_column
        else:
            return TokenType.NUMBER, int(number_str), start_line, start_column
    
    def read_identifier(self) -> Tuple[TokenType, str, int, int]:
        start = self.pos
        start_line = self.line
        start_column = self.column
        
        while self.current_char() and (self.current_char().isalnum() or self.current_char() == '_'):
            self.advance()
        
        identifier = self.source[start:self.pos]
        
        identifier_lower = identifier.lower()
        
        if identifier_lower in KEYWORDS:
            token_type = KEYWORDS[identifier_lower]
            if token_type == TokenType.BOOLEAN:
                value = identifier_lower == 'vrai'
                return TokenType.BOOLEAN, value, start_line, start_column
            
            if identifier_lower == 'sinon':
                self.skip_whitespace(skip_newlines=False)
                if self.current_char() and self.pos + 2 < len(self.source):
                    next_word_match = re.match(r'si\b', self.source[self.pos:], re.IGNORECASE)
                    if next_word_match:
                        for _ in range(len(next_word_match.group())):
                            self.advance()
                        return TokenType.SINON_SI, 'sinon si', start_line, start_column
            
            if identifier_lower == 'tant':
                self.skip_whitespace(skip_newlines=False)
                if self.current_char() and self.pos + 3 < len(self.source):
                    next_word_match = re.match(r'que\b', self.source[self.pos:], re.IGNORECASE)
                    if next_word_match:
                        for _ in range(len(next_word_match.group())):
                            self.advance()
                        return TokenType.TANT_QUE, 'tant que', start_line, start_column
            
            if identifier_lower == 'pour':
                self.skip_whitespace(skip_newlines=False)
                if self.current_char() and self.pos + 6 < len(self.source):
                    next_word_match = re.match(r'chaque\b', self.source[self.pos:], re.IGNORECASE)
                    if next_word_match:
                        for _ in range(len(next_word_match.group())):
                            self.advance()
                        return TokenType.POUR_CHAQUE, 'pour chaque', start_line, start_column
            
            return token_type, identifier, start_line, start_column
        
        return TokenType.IDENTIFIER, identifier, start_line, start_column
    
    def handle_indentation(self, indent_level: int):
        current_indent = self.indent_stack[-1]
        
        if indent_level > current_indent:
            self.indent_stack.append(indent_level)
            self.tokens.append(Token(TokenType.INDENT, indent_level, self.line, 1))
        elif indent_level < current_indent:
            while self.indent_stack and self.indent_stack[-1] > indent_level:
                self.indent_stack.pop()
                self.tokens.append(Token(TokenType.DEDENT, indent_level, self.line, 1))
            
            if not self.indent_stack or self.indent_stack[-1] != indent_level:
                self.error("Indentation incorrecte", code="E103")
    
    def tokenize(self) -> List[Token]:
        while self.pos < len(self.source):
            if self.at_line_start:
                indent_level = 0
                start_pos = self.pos
                
                while self.current_char() in ' \t':
                    if self.current_char() == ' ':
                        indent_level += 1
                    elif self.current_char() == '\t':
                        indent_level += 4
                    self.advance()
                
                if self.current_char() == '#':
                    self.skip_comment()
                    continue
                
                if self.current_char() == '\n':
                    self.advance()
                    continue
                
                if self.current_char() is None:
                    break
                
                self.handle_indentation(indent_level)
                self.at_line_start = False
                continue
            
            char = self.current_char()
            
            if char is None:
                break
            
            if char in ' \t\r':
                self.skip_whitespace(skip_newlines=False)
                continue
            
            if char == '#':
                self.skip_comment()
                continue
            
            if char == '\n':
                self.tokens.append(Token(TokenType.NEWLINE, '\n', self.line, self.column))
                self.advance()
                continue
            
            if char in ('"', "'"):
                string_val, start_line, start_column = self.read_string()
                self.tokens.append(Token(TokenType.STRING, string_val, start_line, start_column))
                continue
            
            if char.isdigit():
                token_type, number_val, start_line, start_column = self.read_number()
                self.tokens.append(Token(token_type, number_val, start_line, start_column))
                continue
            
            if char.isalpha() or char == '_':
                token_type, value, start_line, start_column = self.read_identifier()
                self.tokens.append(Token(token_type, value, start_line, start_column))
                continue
            
            if char == '+':
                self.tokens.append(Token(TokenType.PLUS, '+', self.line, self.column))
                self.advance()
                continue
            
            if char == '-':
                self.tokens.append(Token(TokenType.MINUS, '-', self.line, self.column))
                self.advance()
                continue
            
            if char == '*':
                if self.peek_char() == '*':
                    self.tokens.append(Token(TokenType.POWER, '**', self.line, self.column))
                    self.advance()
                    self.advance()
                else:
                    self.tokens.append(Token(TokenType.MULTIPLY, '*', self.line, self.column))
                    self.advance()
                continue
            
            if char == '/':
                if self.peek_char() == '/':
                    self.tokens.append(Token(TokenType.FLOOR_DIV, '//', self.line, self.column))
                    self.advance()
                    self.advance()
                else:
                    self.tokens.append(Token(TokenType.DIVIDE, '/', self.line, self.column))
                    self.advance()
                continue
            
            if char == '%':
                self.tokens.append(Token(TokenType.MODULO, '%', self.line, self.column))
                self.advance()
                continue
            
            if char == '=':
                if self.peek_char() == '=':
                    self.tokens.append(Token(TokenType.EQUAL, '==', self.line, self.column))
                    self.advance()
                    self.advance()
                else:
                    self.tokens.append(Token(TokenType.ASSIGN, '=', self.line, self.column))
                    self.advance()
                continue
            
            if char == '!':
                if self.peek_char() == '=':
                    self.tokens.append(Token(TokenType.NOT_EQUAL, '!=', self.line, self.column))
                    self.advance()
                    self.advance()
                else:
                    self.error("Opérateur inconnu: !", code="E100")
                continue
            
            if char == '<':
                if self.peek_char() == '=':
                    self.tokens.append(Token(TokenType.LESS_EQUAL, '<=', self.line, self.column))
                    self.advance()
                    self.advance()
                else:
                    self.tokens.append(Token(TokenType.LESS_THAN, '<', self.line, self.column))
                    self.advance()
                continue
            
            if char == '>':
                if self.peek_char() == '=':
                    self.tokens.append(Token(TokenType.GREATER_EQUAL, '>=', self.line, self.column))
                    self.advance()
                    self.advance()
                else:
                    self.tokens.append(Token(TokenType.GREATER_THAN, '>', self.line, self.column))
                    self.advance()
                continue
            
            if char == ':':
                self.tokens.append(Token(TokenType.COLON, ':', self.line, self.column))
                self.advance()
                continue
            
            if char == ',':
                self.tokens.append(Token(TokenType.COMMA, ',', self.line, self.column))
                self.advance()
                continue
            
            if char == '(':
                self.tokens.append(Token(TokenType.LPAREN, '(', self.line, self.column))
                self.advance()
                continue
            
            if char == ')':
                self.tokens.append(Token(TokenType.RPAREN, ')', self.line, self.column))
                self.advance()
                continue
            
            if char == '[':
                self.tokens.append(Token(TokenType.LBRACKET, '[', self.line, self.column))
                self.advance()
                continue
            
            if char == ']':
                self.tokens.append(Token(TokenType.RBRACKET, ']', self.line, self.column))
                self.advance()
                continue
            
            if char == '.':
                self.tokens.append(Token(TokenType.DOT, '.', self.line, self.column))
                self.advance()
                continue
            
            self.error(f"Caractère inattendu: {char}", code="E100")
        
        while len(self.indent_stack) > 1:
            self.indent_stack.pop()
            self.tokens.append(Token(TokenType.DEDENT, 0, self.line, self.column))
        
        self.tokens.append(Token(TokenType.EOF, None, self.line, self.column))
        
        return self.tokens


def tokenize(source: str, filename: str = "<script>") -> List[Token]:
    lexer = Lexer(source, filename)
    return lexer.tokenize()

File: src/lexer.rs
use crate::token::{Token, TokenType, TokenValue};
use crate::error::{ErrorCatalog, MeowLangError};

pub struct Lexer {
    chars: Vec<char>,
    filename: String,
    lines: Vec<String>,
    pos: usize,
    line: usize,
    column: usize,
    tokens: Vec<Token>,
    indent_stack: Vec<usize>,
    at_line_start: bool,
}

impl Lexer {
    pub fn new(source: String, filename: String) -> Self {
        let lines: Vec<String> = source.lines().map(|s| s.to_string()).collect();
        let chars: Vec<char> = source.chars().collect();
        
        Lexer {
            chars,
            filename,
            lines,
            pos: 0,
            line: 1,
            column: 1,
            tokens: Vec::new(),
            indent_stack: vec![0],
            at_line_start: true,
        }
    }
    
    fn current_char(&self) -> Option<char> {
        if self.pos < self.chars.len() {
            Some(self.chars[self.pos])
        } else {
            None
        }
    }
    
    fn peek_char(&self, offset: usize) -> Option<char> {
        if self.pos + offset < self.chars.len() {
            Some(self.chars[self.pos + offset])
        } else {
            None
        }
    }
    
    fn advance(&mut self) {
        if let Some(ch) = self.current_char() {
            if ch == '\n' {
                self.line += 1;
                self.column = 1;
                self.at_line_start = true;
            } else {
                self.column += 1;
            }
            self.pos += 1;
        }
    }
    
    fn skip_whitespace(&mut self, skip_newlines: bool) {
        while let Some(ch) = self.current_char() {
            if ch == ' ' || ch == '\t' || ch == '\r' {
                self.advance();
            } else if skip_newlines && ch == '\n' {
                self.advance();
            } else {
                break;
            }
        }
    }
    
    fn skip_comment(&mut self) {
        if self.current_char() == Some('#') {
            while self.current_char().is_some() && self.current_char() != Some('\n') {
                self.advance();
            }
        }
    }
    
    fn read_string(&mut self) -> Result<String, MeowLangError> {
        let start_line = self.line;
        let start_column = self.column;
        let quote_char = self.current_char().unwrap();
        self.advance();
        
        let mut result = String::new();
        
        while self.current_char().is_some() && self.current_char() != Some(quote_char) {
            if self.current_char() == Some('\\') {
                self.advance();
                if let Some(ch) = self.current_char() {
                    let escaped = match ch {
                        'n' => '\n',
                        't' => '\t',
                        'r' => '\r',
                        '\\' => '\\',
                        '"' => '"',
                        '\'' => '\'',
                        _ => ch,
                    };
                    result.push(escaped);
                    self.advance();
                }
            } else {
                result.push(self.current_char().unwrap());
                self.advance();
            }
        }
        
        if self.current_char().is_none() {
            return Err(MeowLangError::new(
                ErrorCatalog::get("E101"),
                self.filename.clone(),
                start_line,
                start_column,
            ).with_context(&self.lines));
        }
        
        self.advance();
        Ok(result)
    }
    
    fn read_number(&mut self) -> (f64, bool) {
        let start = self.pos;
        let mut has_dot = false;
        
        while let Some(ch) = self.current_char() {
            if ch.is_ascii_digit() {
                self.advance();
            } else if ch == '.' && !has_dot {
                if let Some(next) = self.peek_char(1) {
                    if next.is_ascii_digit() {
                        has_dot = true;
                        self.advance();
                    } else {
                        break;
                    }
                } else {
                    break;
                }
            } else {
                break;
            }
        }
        
        let number_str: String = self.chars[start..self.pos].iter().collect();
        let number = number_str.parse::<f64>().unwrap_or(0.0);
        (number, has_dot)
    }
    
    fn read_identifier(&mut self) -> String {
        let start = self.pos;
        
        while let Some(ch) = self.current_char() {
            if ch.is_alphanumeric() || ch == '_' {
                self.advance();
            } else {
                break;
            }
        }
        
        self.chars[start..self.pos].iter().collect()
    }
    
    fn get_keyword_token(&mut self, identifier: &str, line: usize, column: usize) -> Token {
        let lower = identifier.to_lowercase();
        
        match lower.as_str() {
            "miaou" => Token::simple(TokenType::Miaou, line, column),
            "meow" => Token::simple(TokenType::Meow, line, column),
            "ecrire" => Token::simple(TokenType::Ecrire, line, column),
            "demander" => Token::simple(TokenType::Demander, line, column),
            "si" => Token::simple(TokenType::Si, line, column),
            "alors" => Token::simple(TokenType::Alors, line, column),
            "sinon" => {
                self.skip_whitespace(false);
                if let Some(ch) = self.current_char() {
                    if ch == 's' {
                        let next_word = self.peek_identifier();
                        if next_word.to_lowercase() == "si" {
                            self.read_identifier();
                            return Token::simple(TokenType::SinonSi, line, column);
                        }
                    }
                }
                Token::simple(TokenType::Sinon, line, column)
            },
            "repeter" => Token::simple(TokenType::Repeter, line, column),
            "fois" => Token::simple(TokenType::Fois, line, column),
            "tant" => {
                self.skip_whitespace(false);
                if let Some(ch) = self.current_char() {
                    if ch == 'q' {
                        let next_word = self.peek_identifier();
                        if next_word.to_lowercase() == "que" {
                            self.read_identifier();
                            return Token::simple(TokenType::TantQue, line, column);
                        }
                    }
                }
                Token::new(TokenType::Identifier, TokenValue::String(identifier.to_string()), line, column)
            },
            "pour" => {
                self.skip_whitespace(false);
                if let Some(ch) = self.current_char() {
                    if ch == 'c' {
                        let next_word = self.peek_identifier();
                        if next_word.to_lowercase() == "chaque" {
                            self.read_identifier();
                            return Token::simple(TokenType::PourChaque, line, column);
                        }
                    }
                }
                Token::new(TokenType::Identifier, TokenValue::String(identifier.to_string()), line, column)
            },
            "dans" => Token::simple(TokenType::Dans, line, column),
            "compteur" => Token::simple(TokenType::Compteur, line, column),
            "fonction" => Token::simple(TokenType::Fonction, line, column),
            "retour" => Token::simple(TokenType::Retour, line, column),
            "liste" => Token::simple(TokenType::Liste, line, column),
            "dictionnaire" => Token::simple(TokenType::Dictionnaire, line, column),
            "essayer" => Token::simple(TokenType::Essayer, line, column),
            "sauf" => Token::simple(TokenType::Sauf, line, column),
            "erreur" => Token::simple(TokenType::Erreur, line, column),
            "importer" => Token::simple(TokenType::Importer, line, column),
            "minuscule" => Token::simple(TokenType::Minuscule, line, column),
            "majuscule" => Token::simple(TokenType::Majuscule, line, column),
            "longueur" => Token::simple(TokenType::Longueur, line, column),
            "remplacer" => Token::simple(TokenType::Remplacer, line, column),
            "contient" => Token::simple(TokenType::Contient, line, column),
            "aleatoire" => Token::simple(TokenType::Aleatoire, line, column),
            "sqrt" => Token::simple(TokenType::Sqrt, line, column),
            "abs" => Token::simple(TokenType::Abs, line, column),
            "round" => Token::simple(TokenType::Round, line, column),
            "floor" => Token::simple(TokenType::Floor, line, column),
            "ceil" => Token::simple(TokenType::Ceil, line, column),
            "ouvrir" => Token::simple(TokenType::Ouvrir, line, column),
            "lire" => Token::simple(TokenType::Lire, line, column),
            "fermer" => Token::simple(TokenType::Fermer, line, column),
            "attendre" => Token::simple(TokenType::Attendre, line, column),
            "vrai" => Token::new(TokenType::Boolean, TokenValue::Boolean(true), line, column),
            "faux" => Token::new(TokenType::Boolean, TokenValue::Boolean(false), line, column),
            "et" => Token::simple(TokenType::Et, line, column),
            "ou" => Token::simple(TokenType::Ou, line, column),
            "non" => Token::simple(TokenType::Non, line, column),
            "a" => Token::simple(TokenType::A, line, column),
            _ => Token::new(TokenType::Identifier, TokenValue::String(identifier.to_string()), line, column),
        }
    }
    
    fn peek_identifier(&self) -> String {
        let mut pos = self.pos;
        let mut result = String::new();
        
        while pos < self.chars.len() {
            let ch = self.chars[pos];
            if ch.is_alphanumeric() || ch == '_' {
                result.push(ch);
                pos += 1;
            } else {
                break;
            }
        }
        
        result
    }
    
    fn handle_indentation(&mut self, indent_level: usize) {
        let current_indent = *self.indent_stack.last().unwrap();
        
        if indent_level > current_indent {
            self.indent_stack.push(indent_level);
            self.tokens.push(Token::new(
                TokenType::Indent,
                TokenValue::Indent(indent_level),
                self.line,
                1,
            ));
        } else if indent_level < current_indent {
            while !self.indent_stack.is_empty() && *self.indent_stack.last().unwrap() > indent_level {
                self.indent_stack.pop();
                self.tokens.push(Token::new(
                    TokenType::Dedent,
                    TokenValue::Indent(indent_level),
                    self.line,
                    1,
                ));
            }
        }
    }
    
    pub fn tokenize(&mut self) -> Result<Vec<Token>, MeowLangError> {
        while self.pos < self.chars.len() {
            if self.at_line_start {
                let mut indent_level = 0;
                
                while let Some(ch) = self.current_char() {
                    if ch == ' ' {
                        indent_level += 1;
                        self.advance();
                    } else if ch == '\t' {
                        indent_level += 4;
                        self.advance();
                    } else {
                        break;
                    }
                }
                
                if self.current_char() == Some('#') {
                    self.skip_comment();
                    continue;
                }
                
                if self.current_char() == Some('\n') {
                    self.advance();
                    continue;
                }
                
                if self.current_char().is_none() {
                    break;
                }
                
                self.handle_indentation(indent_level);
                self.at_line_start = false;
                continue;
            }
            
            let ch = match self.current_char() {
                Some(c) => c,
                None => break,
            };
            
            if ch == ' ' || ch == '\t' || ch == '\r' {
                self.skip_whitespace(false);
                continue;
            }
            
            if ch == '#' {
                self.skip_comment();
                continue;
            }
            
            if ch == '\n' {
                self.tokens.push(Token::simple(TokenType::Newline, self.line, self.column));
                self.advance();
                continue;
            }
            
            if ch == '"' || ch == '\'' {
                let line = self.line;
                let column = self.column;
                let string_val = self.read_string()?;
                self.tokens.push(Token::new(
                    TokenType::String,
                    TokenValue::String(string_val),
                    line,
                    column,
                ));
                continue;
            }
            
            if ch.is_ascii_digit() {
                let line = self.line;
                let column = self.column;
                let (number, has_dot) = self.read_number();
                
                if has_dot {
                    self.tokens.push(Token::new(TokenType::Number, TokenValue::Number(number), line, column));
                } else {
                    self.tokens.push(Token::new(TokenType::Number, TokenValue::Integer(number as i64), line, column));
                }
                continue;
            }
            
            if ch.is_alphabetic() || ch == '_' {
                let line = self.line;
                let column = self.column;
                let identifier = self.read_identifier();
                let token = self.get_keyword_token(&identifier, line, column);
                self.tokens.push(token);
                continue;
            }
            
            let line = self.line;
            let column = self.column;
            
            match ch {
                '+' => {
                    self.advance();
                    self.tokens.push(Token::simple(TokenType::Plus, line, column));
                },
                '-' => {
                    self.advance();
                    self.tokens.push(Token::simple(TokenType::Minus, line, column));
                },
                '*' => {
                    self.advance();
                    if self.current_char() == Some('*') {
                        self.advance();
                        self.tokens.push(Token::simple(TokenType::Power, line, column));
                    } else {
                        self.tokens.push(Token::simple(TokenType::Multiply, line, column));
                    }
                },
                '/' => {
                    self.advance();
                    if self.current_char() == Some('/') {
                        self.advance();
                        self.tokens.push(Token::simple(TokenType::FloorDiv, line, column));
                    } else {
                        self.tokens.push(Token::simple(TokenType::Divide, line, column));
                    }
                },
                '%' => {
                    self.advance();
                    self.tokens.push(Token::simple(TokenType::Modulo, line, column));
                },
                '=' => {
                    self.advance();
                    if self.current_char() == Some('=') {
                        self.advance();
                        self.tokens.push(Token::simple(TokenType::Equal, line, column));
                    } else {
                        self.tokens.push(Token::simple(TokenType::Assign, line, column));
                    }
                },
                '!' => {
                    self.advance();
                    if self.current_char() == Some('=') {
                        self.advance();
                        self.tokens.push(Token::simple(TokenType::NotEqual, line, column));
                    }
                },
                '<' => {
                    self.advance();
                    if self.current_char() == Some('=') {
                        self.advance();
                        self.tokens.push(Token::simple(TokenType::LessEqual, line, column));
                    } else {
                        self.tokens.push(Token::simple(TokenType::LessThan, line, column));
                    }
                },
                '>' => {
                    self.advance();
                    if self.current_char() == Some('=') {
                        self.advance();
                        self.tokens.push(Token::simple(TokenType::GreaterEqual, line, column));
                    } else {
                        self.tokens.push(Token::simple(TokenType::GreaterThan, line, column));
                    }
                },
                ':' => {
                    self.advance();
                    self.tokens.push(Token::simple(TokenType::Colon, line, column));
                },
                ',' => {
                    self.advance();
                    self.tokens.push(Token::simple(TokenType::Comma, line, column));
                },
                '(' => {
                    self.advance();
                    self.tokens.push(Token::simple(TokenType::LParen, line, column));
                },
                ')' => {
                    self.advance();
                    self.tokens.push(Token::simple(TokenType::RParen, line, column));
                },
                '[' => {
                    self.advance();
                    self.tokens.push(Token::simple(TokenType::LBracket, line, column));
                },
                ']' => {
                    self.advance();
                    self.tokens.push(Token::simple(TokenType::RBracket, line, column));
                },
                '.' => {
                    self.advance();
                    self.tokens.push(Token::simple(TokenType::Dot, line, column));
                },
                _ => {
                    return Err(MeowLangError::new(
                        ErrorCatalog::get("E100"),
                        self.filename.clone(),
                        line,
                        column,
                    ).with_instruction(ch.to_string()).with_context(&self.lines));
                }
            }
        }
        
        while self.indent_stack.len() > 1 {
            self.indent_stack.pop();
            self.tokens.push(Token::simple(TokenType::Dedent, self.line, self.column));
        }
        
        self.tokens.push(Token::simple(TokenType::Eof, self.line, self.column));
        
        Ok(self.tokens.clone())
    }
}
